fbpic/boundaries/particle_buffer_handling.py

```diff
@@ -196,18 +196,28 @@ def remove_particles_gpu(species, fld, nguard, left_proc, right_proc):
     prefix_sum = species.prefix_sum
     Nz = fld.Nz
     Nr = fld.Nr
-    i_min = prefix_sum.getitem( (nguard+fld.prefix_sum_shift)*Nr )
-    i_max = prefix_sum.getitem( (Nz-nguard+fld.prefix_sum_shift)*Nr - 1 )
-    # For the open boundaries, only the particles in the outermost
-    # half of the guard cells are removed
+    # Find the z index of the first cell for which particles are kept
     if left_proc is None:
-        # Find the index in z below which particles are removed
+        # Open boundary: particles in outermost half of guard cells are removed
         iz_min = max( int(nguard/2) + fld.prefix_sum_shift, 0 )
-        i_min = prefix_sum.getitem( iz_min * Nr )
```

Comment:
The error in the previous version of the code is that I forgot to add -1 in the above line, i.e. I have
```
i_min = prefix_sum.getitem( iz_min * Nr )
```
instead of 
```
i_min = prefix_sum.getitem( iz_min * Nr - 1 )
```
This meant that, in addition to the particles of the left guard cells, we also passed the particles of the **first cell of the valid physical domain** to the left proc -- which is invalid. Note that the calculation of `i_max` was correct. I don't know why I did it for `i_max` and not `i_min`.

Anyway, once I corrected this bug, I also realized this whole part would be slightly more readable if reorganize it. Sorry if that makes the PR less readable. But the real only change is the -1, the rest is refactoring.